### ISSUE TITLE
Split `Client` into connection & `start_request` handles.

### ISSUE THREAD

**Issue by carllerche** [carllerche/h2#104]
Conflating the handle the advances the connection state with the handle that is used to start requests is limiting. These handles should be split into two types.

## Open questions

* What should the two types be named
* Should the `start_request` handle impl `Clone`.

Specifically, if `StartRequest` (naming tbd) impls `Clone` this would allow streams to be initialized from multiple tasks/threads, which seems like a good feature. The difficulty lies in coordinating access to a single resource across tasks. When the concurrent stream limit is reached, `StartRequest::poll_ready` calls would race to get access to stream IDs. The solution to this could be to model how `mpsc::Sender` behaves when cloned.

**Comment by seanmonstar** [comment]
- `Connection` and `Client`?
- Cloning the sender part seems nice. Is there any concern about multi-threaded access and the `Mutex` used internally to `Streams`?

**Comment by carllerche** [comment]
`Connection` and `Client` seems fine.

The issue w/ cloning isn't so much the required synchronization (there already is a mutex), but the general problem w/ having async Sink type of behavior that push to a single resource. There could be 1000 outstanding `Client` handles and only 1 available `StreamId`. You want to avoid thundering herds when notifying tasks.

The  `mpsc::Sender` strategy is that each handle has a guaranteed buffer slot. IMO that would work fine for this as well.

**Comment by seanmonstar** [comment]
That makes sense.

Actually, to clone the `Connection`, what would you expect for synchronization? Putting an `Arc<Mutex<proto::Connection>>` in both `client::{Client, Connection}`?

**Comment by seanmonstar** [comment]
An alternative to having to deal with thread synchronization is to push the cost out of h2, and let users who need it pay the cost. For example:

Single-threaded:

```rust
let (conn, client) = ...;
handle.spawn(conn.map_err(|_| ()));
let fut = client.send_request(req).and_then(|resp| {
    // ...
}).map_err(|_| ());
handle.spawn(fut);
```

Multi-threaded:

```rust
let (conn, client) = ...;
handle.spawn(conn.map_err(|_| ()));
let (tx, rx) = mpsc::channel();

handle.spawn(rx.for_each(move |(req, oneshot)| {
    handle.spawn(client.send_request(req)
        .then(|res| oneshot.send(res))
        .map_err(|_| ()))
    Ok(())
});
// give clones of `tx` out like candy
```

**Comment by carllerche** [comment]
In this case, you wouldn't clone `Connection` (since it is only used to advanced the connection state).

Re: `Client::clone` the reason to have h2 support it directly is that h2 already internally has all the necessary storage / machinery to handle it, so you could avoid the oneshot in your example.

If a request is sent when there are no available stream IDs, it would queue up the request frame internally and wait until a StreamId becomes available.

**Comment by seanmonstar** [comment]
Ok, I traced the methods the `Client` side would have and see what you mean. It can just clone the `Streams` out of the `Connection`.

Interestingly, this means `Client` doesn't need the `T: AsyncRead + AsyncWrite` generic. Does it seem nicer to remove that, and just have `Client<B>`, or push it into a `PhatomData`, so you can still know what is driving the `Client`?

**Comment by olix0r** [comment]
i'd probably err towards omitting `T` from client if it isn't strictly necessary?

**Comment by carllerche** [comment]
It seems nice to remove the `AsyncRead + AsyncWrite` generic from `Client`, especially since `Stream` / `Body` / etc.. don't have it.

**Comment by seanmonstar** [comment]
Merged with #107.
